juju/juju
# Bind Dqlite to a local-cloud address if possible at bootstrap

JUJU-4611

To-date we have bound Dqlite to the loopback address (127.0.0.1) at bootstrap, and rebound it to a local-cloud address when entering HA.

What has been observed, particularly in testing by solutions QA, is that the Dqlite rebind step and the attendant churn is causing a period where the controller is inaccessible.

In this change, we will bind Dqlite to a local-cloud address at bootstrap, if we can determine that the address is unique in that scope. This means there is no disruption at all when entering HA, and HA is established much sooner.

If there is no unique local-cloud address determined, we simply fall back to the prior behaviour of using the loopback IP, and rebinding later based on the peer-grouper's address determination.

## QA steps

- Bootstrap to LXD (the default profile/bridge should ensure a unique address).
- `juju ssh -m controller 0`.
- `sudo cat /var/lib/juju/dqlite/info.yaml` will show we're not bound to the loopback.
- `juju enable-ha`.
- `juju debug-log -m controller` will indicate entry into HA _without_ disconnecting as it did before.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/2015371
